src/fhirdefs/load.ts

```diff
@@ -1,29 +1,118 @@
-    const files = [
-      `${__dirname}/fhir-${fhirVersion}/extension-definitions.json`,
-      `${__dirname}/fhir-${fhirVersion}/profiles-resources.json`,
-      `${__dirname}/fhir-${fhirVersion}/profiles-types.json`,
-      `${__dirname}/fhir-${fhirVersion}/profiles-others.json`,
-      `${__dirname}/fhir-${fhirVersion}/valuesets.json`
-    ];
-    for (const file of files) {
-      const definitions = JSON.parse(fs.readFileSync(file, 'utf-8'));
-      for (const entry of definitions.entry) {
-        result.add(entry.resource);
+/**
+ * Loads a dependency from user FHIR cache or from online
+ * @param {string} packageName - The name of the package to load
+ * @param {string} version - The version of the package to load
+ * @param {FHIRDefinitions} FHIRDefs - The FHIRDefinitions to load the dependencies into
+ * @param {string} cachePath - The path to load the package into
+ * @returns {Promise<FHIRDefinitions>} the loaded FHIRDefs
+ * @throws {PackageLoadError} when the desired package can't be loaded
+ */
+export async function loadDependency(
+  packageName: string,
+  version: string,
+  FHIRDefs: FHIRDefinitions,
+  cachePath: string = path.join(os.homedir(), '.fhir', 'packages')
+): Promise<FHIRDefinitions> {
+  const fullPackageName = `${packageName}#${version}`;
+  const loadPath = path.join(cachePath, fullPackageName, 'package');
+  let loadedPackage: string;
+  if (version != 'current') {
+    loadedPackage = loadFromPath(loadPath, fullPackageName, FHIRDefs);
+  }
+  if (!loadedPackage) {
+    let packageUrl: string;
+    if (version === 'dev') {
+      // Dev packages must be present in local FHIR cache
+      throw new DevPackageLoadError(fullPackageName);
+    } else if (version === 'current') {
+      // Current packages need to be loaded using build.fhir.org
+      const baseUrl = 'http://build.fhir.org/ig';
+      const res = await rp.get({ uri: `${baseUrl}/qas.json`, json: true });
+      // Find matching packages and sort by date to get the most recent
+      let newestPackage;
+      if (res && res.length > 0) {
+        const matchingPackages = res.filter((p: any) => p['package-id'] === packageName);
+        newestPackage = matchingPackages.sort((p1: any, p2: any) => {
+          return Date.parse(p2['date']) - Date.parse(p1['date']);
+        })[0];
+      }
+      if (newestPackage) {
+        // Current packages are stored at build.fhir.org
+        packageUrl = `${baseUrl}/${newestPackage.repo}/package.tgz`;
+      } else {
+        throw new CurrentPackageLoadError(fullPackageName);
       }
+    } else {
+      // All non-current packages are stored at packages.fhir.org
+      packageUrl = `http://packages.fhir.org/${packageName}/${version}`;
     }
+    // Create a temporary file and write the package to there
+    temp.track();
+    const tempFile = temp.openSync();
+    const targetDirectory = path.join(cachePath, fullPackageName);
+    const res = await rp.get({ uri: packageUrl, encoding: null });
```

Comment:
There's an additional error case that I think should be handled here: if a dependency specifies a version that doesn't exist, the promise here will be rejected, which currently causes an `UnhandledPromiseRejectionWarning` to appear in the output. When this `get` fails, it would be useful to show a message based on the HTTP response code: 404 probably means that the user specified a package and version combination that doesn't exist, 5XX probably just means the server is down.